homeassistant/components/acmeda/config_flow.py

```diff
@@ -0,0 +1,119 @@
+import asyncio
+from typing import Dict, Optional
+
+import aiopulse
+import async_timeout
+import voluptuous as vol
+
+from homeassistant import config_entries
+
+from .const import DOMAIN, LOGGER  # pylint: disable=unused-import
+from .errors import CannotConnect
+
+
+class AcmedaFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Acmeda config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self.hub: Optional[aiopulse.Hub] = None
+        self.discovered_hubs: Optional[Dict[str, aiopulse.Hub]] = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        if (
+            user_input is not None
+            and self.discovered_hubs is not None
+            # pylint: disable=unsupported-membership-test
+            and user_input["id"] in self.discovered_hubs
+        ):
+            # pylint: disable=unsubscriptable-object
+            self.hub = self.discovered_hubs[user_input["id"]]
+            await self.async_set_unique_id(self.hub.id, raise_on_progress=False)
+            # We pass user input to link so it will attempt to link right away
+            return await self.async_step_link()
+
+        hubs = []
+        try:
+            with async_timeout.timeout(5):
+                async for hub in aiopulse.Hub.discover():
+                    hubs.append(hub)
+        except asyncio.TimeoutError:
+            if len(hubs) == 0:
+                return self.async_abort(reason="discover_timeout")
+
+        if not hubs:
+            return self.async_abort(reason="no_hubs")
+
+        # Find already configured hosts
+        already_configured = {
+            entry.data["host"] for entry in self._async_current_entries()
+        }
+        hubs = [hub for hub in hubs if hub.host not in already_configured]
+
+        if not hubs:
+            return self.async_abort(reason="all_configured")
+
+        if len(hubs) == 1:
+            self.hub = hubs[0]
+            await self.async_set_unique_id(self.hub.id, raise_on_progress=False)
+            return await self.async_step_link()
+
+        self.discovered_hubs = {hub.id: hub for hub in hubs}
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
+                {vol.Required("id"): vol.In({hub.id: hub.host for hub in hubs})}
+            ),
+        )
+
+    async def async_step_link(self, user_input=None):
+        """Attempt to link with the Acmeda Hub."""
+        hub = self.hub
+        assert hub is not None
+        errors = {}
+
+        try:
+            # Can happen if we come from import.
+            if self.unique_id is None:
+                await self.async_set_unique_id(hub.id, raise_on_progress=False)
+
+            if hub.id:
+                title = f"{hub.id} ({hub.host})"
+            else:
```

Comment:
If there's not hub id we can't set unique_id for the config entry?

Reply:
The challenge I found was that if the hub was configured in the yaml then we won't know the hub.id until it connects for the first time. Should I mandate that that id is provided in the yaml and reject the connection if the specified ID doesn't match what the device reports?

Reply:
We should check the hub id after connection also for the import flow and abort if that id is already set up.

Reply:
So should it try and connect during the import flow to obtain the hub id?
If so, what should it do if it can't currently connect?

The current approach is asynchronous such that the config flow can complete even before the first successful connection is made. This means that on startup devices that aren't responding are still shown in the integrations and can be removed by the user.

Reply:
We should always check that we can connect before creating a config entry. If we can't connect during the import flow, log error and abort.

Reply:
I implemented doing a connect before creating a config entry but found that this stalled the starting of home assistant. Noting that YAML configuration is deprecated I opted to remove the YAML config import support.